libs/elasticsearch/langchain_elasticsearch/cache.py

```diff
@@ -0,0 +1,191 @@
+                self._es_client = create_elasticsearch_client(
+                    url=es_url,
+                    cloud_id=es_cloud_id,
+                    api_key=es_api_key,
+                    username=es_user,
+                    password=es_password,
+                    params=es_params,
+                )
+            except Exception as err:
+                logger.error(f"Error connecting to Elasticsearch: {err}")
+                raise err
+        else:
+            raise ValueError(
+                """Either provide a pre-existing Elasticsearch connection, \
+                or valid credentials for creating a new connection."""
+            )
+        self._es_index = es_index
+        self._store_input = store_input
+        self._store_timestamp = store_timestamp
+        self._store_input_params = store_input_params
+        self._metadata = metadata
+        self._manage_index()
+
+    def _manage_index(self) -> None:
+        """Write or update an index or alias according to the default mapping"""
+        self._is_alias = False
+        if self._es_client.indices.exists_alias(name=self._es_index):
+            self._is_alias = True
+        elif not self._es_client.indices.exists(index=self._es_index):
+            logger.debug(f"Creating new Elasticsearch index: {self._es_index}")
+            self._es_client.indices.create(index=self._es_index, body=self.mapping)
+            return
+        self._es_client.indices.put_mapping(
+            index=self._es_index, body=self.mapping["mappings"]
+        )
+
+    @cached_property
+    def mapping(self) -> Dict[str, Any]:
+        """Get the default mapping for the index."""
+        return {
+            "mappings": {
+                "properties": {
+                    "llm_output": {"type": "text", "index": False},
+                    "llm_params": {"type": "text", "index": False},
+                    "llm_input": {"type": "text", "index": False},
+                    "metadata": {"type": "object"},
+                    "timestamp": {"type": "date"},
+                }
+            }
+        }
+
+    @staticmethod
+    def _key(prompt: str, llm_string: str) -> str:
+        """Generate a key for the cache store."""
+        return hashlib.md5((prompt + llm_string).encode()).hexdigest()
+
+    def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
+        """Look up based on prompt and llm_string."""
+        cache_key = self._key(prompt, llm_string)
+        if self._is_alias:
+            result = self._es_client.search(
+                index=self._es_index,
+                body={"query": {"term": {"_id": cache_key}}},
+                source_includes=["llm_output"],
+            )
+            if result["hits"]["total"]["value"] > 0:
+                # get the record from the latest index,
+                # assuming lexicographic order is chronological
+                record = max(result["hits"]["hits"], key=itemgetter("_index"))
```

Comment:
Can it ever happen that we get multiple results for the same `cache_key`? I think not. But if yes, this will not give us the "latest" entry necessarily. Here we are looking for max on index name but we only have one index name when using this cache. So I think we cannot assume anything about the ordering. You might and to consider to [sort](https://www.elastic.co/guide/en/elasticsearch/reference/current/sort-search-results.html) the output by timestamp.

Reply:
I have changed it to a search sort by timestamp. I applied a too strong assumption on the indices. However, the case where two identical records are on different indices could happen if someone plays with the alias (i.e. putting back old indices under an alias, for example for "rollback" of the cached data)

Reply:
Makes sense what you are saying. We should consider now to always store the timestamp though (remove the `store_timestamp` option) because the lookup is now relying on it in the case of aliases. If you feel strongly about keeping it, it would be good to have an error message that explains that `store_timestamp=False` cannot be used in combination with aliases.

Reply:
sorry, you are right, I think the best solution (the one I have committed) is to "force" the timestamp in the document, as we do for the `llm_output`. It is an essential information for any cache retention policy one would implement